torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,143 @@
+        ValueError: if ``embedding_dim`` is not an integer multiple of ``len(shape)``
+
+    Inputs:
+        x (Optional[torch.Tensor]): flattened input data, e.g., ``(batch, time * height * width, embedding_dim)``
+            if ``dim=-1``.
+        decode_step (Optional[int]): position of the data that requires decoding.
+    """
+
+    def __init__(
+        self, shape: torch.Size, embedding_dim: int, dim: Optional[int] = -1
+    ) -> None:
+        super().__init__()
+        if dim not in [-1, 1]:
+            raise ValueError(
+                f"Only first or last for the embedding dim is supported but got {dim}"
+            )
+        if embedding_dim % len(shape) != 0:
+            raise ValueError(
+                f"Embedding dim {embedding_dim} modulo len(shape) {len(shape)} is not zero"
+            )
+
+        self.shape = shape
+        self.n_dim = n_dim = len(shape)
+        self.embedding_dim = embedding_dim
+        self.dim = dim
+
+        self.embedding = nn.ParameterDict(
+            {
+                f"d_{i}": nn.Parameter(
+                    torch.randn(shape[i], embedding_dim // n_dim) * 0.01
+                    if dim == -1
+                    else torch.randn(embedding_dim // n_dim, shape[i]) * 0.01
+                )
+                for i in range(n_dim)
+            }
+        )
+
+    @property
+    def seq_len(self) -> int:
+        """Dimension of flattened data, e.g., time * height * width"""
+        return np.prod(self.shape)
+
+    @property
+    def decode_idxs(self):
+        """Indices along the dims of data, e.g., ``(time, height, width)``."""
+        return list(itertools.product(*[range(s) for s in self.shape]))
+
+    def _broadcast(self, emb: torch.Tensor, i: int) -> torch.Tensor:
+        """Broadcasts the ``i``-th embedding matrix ``(self.shape[i], self.embedding_dim // n_dim)`` along the other
+        dims of ``self.shape``. The embedding dim is kept at ``self.dim``.
+
+        For example::
+
+            >>> pos_emb = BroadcastPositionEmbedding(shape=(2, 4), embedding_dim=6)
+            >>> print(pos_emb.embedding["d_0"]).shape
+            torch.Size([2, 3])
+            >>> out = pos_emb._broadcast(torch.tensor([[0, 0, 0],[0, 0, 1]]), i=0)
+            >>> print(out)
+            tensor([[[[0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]],
+                    [[0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1]]]])
+
+        The input is broadcasted along dim ``4`` since it's the ``0``-th embedding constructed w.r.t dim ``2``.
+        """
+        if self.dim == -1:
+            # (1, 1, ..., 1, self.shape[i], 1, ..., -1)
+            emb = emb.view(
+                1, *((1,) * i), self.shape[i], *((1,) * (self.n_dim - i - 1)), -1
+            )
+            # (1, *self.shape, -1)
+            emb = emb.expand(1, *self.shape, -1)
+        else:
+            emb = emb.view(
+                1, -1, *((1,) * i), self.shape[i], *((1,) * (self.n_dim - i - 1))
```

Comment:
Another readability thing: mixing the uses of * here makes me feel like I am reading C++ (and not in a good way 😅 )